File: packages/cli/src/commands/diff-base.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

import type { Command } from 'commander';

import { applyPatch, patchFromYaml } from '@smokepingconf/core';

import { resolveBase } from '../base-resolver.js';

interface DiffBaseOpts {
  base?: string;
  baseUrl?: string;
  onDrift: 'ignore' | 'warn' | 'error';
}

export function registerDiffBase(program: Command): void {
  program
    .command('diff-base <patch>')
    .description('report drift between a patch and its resolved base catalogue')
    .option('-b, --base <file>', 'local catalog.json path (overrides --base-url/bundled)')
    .option('-u, --base-url <url>', 'fetch catalog.json over HTTP(S)')
    .option(
      '--on-drift <mode>',
      'exit non-zero if drift is detected: ignore | warn | error',
      'warn',
    )
    .action(async (patchPath: string, opts: DiffBaseOpts) => {
      process.exit(await runDiffBase(patchPath, opts));
    });
}

export async function runDiffBase(patchPath: string, opts: DiffBaseOpts): Promise<number> {
  const base = await resolveBase(opts);
  const patchText = readFileSync(resolve(process.cwd(), patchPath), 'utf8');
  const patch = patchFromYaml(patchText);
  const { drift } = applyPatch(patch, base);

  const cleanReport = !drift.baseMismatch && drift.missingPaths.length === 0;
  if (cleanReport) {
    process.stdout.write('no drift — patch applies cleanly against the current base\n');
    return 0;
  }

  if (drift.baseMismatch) {
    const p = drift.baseMismatch.patch;
    const a = drift.baseMismatch.actual;
    process.stdout.write('baseVersion mismatch:\n');
    process.stdout.write(`  patch pinned: ${p.date} @ ${p.sha}\n`);
    process.stdout.write(`  current base: ${a ? `${a.date} @ ${a.sha}` : '(no version stamp)'}\n`);
  }
  if (drift.missingPaths.length > 0) {
    process.stdout.write(
      `paths referenced by patch but missing from base (${drift.missingPaths.length}):\n`,
    );
    for (const p of drift.missingPaths) process.stdout.write(`  - ${p}\n`);
  }

  return opts.onDrift === 'error' ? 1 : 0;
}

File: packages/cli-rs/src/commands/diff_base.rs
use std::fs;
use std::path::Path;

use crate::base_resolver::resolve_base;
use crate::patch::{apply_patch, patch_from_yaml};

pub fn run_diff_base(
    patch_path: &str,
    base: Option<&str>,
    base_url: Option<&str>,
    on_drift: &str,
) -> i32 {
    let catalog = match resolve_base(base, base_url) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("smokepingconf: {}", e);
            return 1;
        }
    };

    let abs = if Path::new(patch_path).is_absolute() {
        patch_path.to_string()
    } else {
        std::env::current_dir()
            .unwrap()
            .join(patch_path)
            .to_string_lossy()
            .to_string()
    };
    let patch_text = match fs::read_to_string(&abs) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("smokepingconf: {}: {}", patch_path, e);
            return 1;
        }
    };

    let patch = match patch_from_yaml(&patch_text) {
        Ok(p) => p,
        Err(e) => {
            eprintln!("smokepingconf: {}", e);
            return 1;
        }
    };

    let (_, drift) = apply_patch(&patch, &catalog);

    let clean = drift.base_mismatch.is_none() && drift.missing_paths.is_empty();
    if clean {
        println!("no drift — patch applies cleanly against the current base");
        return 0;
    }

    if let Some(mm) = &drift.base_mismatch {
        let p = &mm.patch;
        println!("baseVersion mismatch:");
        println!("  patch pinned: {} @ {}", p.date, p.sha);
        if let Some(a) = &mm.actual {
            println!("  current base: {} @ {}", a.date, a.sha);
        } else {
            println!("  current base: (no version stamp)");
        }
    }

    if !drift.missing_paths.is_empty() {
        println!(
            "paths referenced by patch but missing from base ({}):",
            drift.missing_paths.len()
        );
        for p in &drift.missing_paths {
            println!("  - {}", p);
        }
    }

    if on_drift == "error" { 1 } else { 0 }
}
